### PR TITLE
added a feature for private repositories

### PR DESCRIPTION
**Pull request by Mebus**

### CODE REVIEW

**Comment by wojas** [comment]
Great feature!

You could reduce code duplication by doing this check in `handlers.AuthHandler`.

**Comment by wojas** [comment]
And you can fix the CI build by running `goimports -w *.go` to reformat the code.

**Comment by Mebus** [comment]
I tried to do it in handlers.AuthHandler, but I couldn't get the repo name there.

**Comment by zcalusic** [comment]
Hello @Mebus and thanks for your contribution.

Somehow, I share the opinion with @wojas that there would be much less repetition if this could be solved via AuthHandler, 'cause it basically belongs there (access control). Without much time to investigate right now, I suspect it's the problem of passing info between two handlers?

Also, there should be a piece of documentation explaining what the switch does, before it's merged.

Let's see if it can be solved in a cleaner way... @wojas, if you have any ideas, you're welcome.

**Comment by wojas** [comment]
This would work:
```go
// isUserPath checks if a request path is accessible by the user when using
// private repositories.
func isUserPath(username, path string) bool {
	prefix := "/" + username
	if !strings.HasPrefix(path, prefix) {
		return false
	}
	return len(path) == len(prefix) || path[len(prefix)] == '/'
}

// AuthHandler wraps h with a http.HandlerFunc that performs basic authentication against the user/passwords pairs
// stored in f, performs permission checks and returns the http.HandlerFunc.
func AuthHandler(f *HtpasswdFile, h http.Handler) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		username, password, ok := r.BasicAuth()
		if !ok || !f.Validate(username, password) {
			http.Error(w, http.StatusText(http.StatusUnauthorized), http.StatusUnauthorized)
			return
		}
		if Config.PrivateRepos && !isUserPath(username, r.URL.Path) {
			http.Error(w, http.StatusText(http.StatusUnauthorized), http.StatusUnauthorized)
			return
		}
		h.ServeHTTP(w, r)
	}
}
```

Test:
```go
func TestIsUserPath(t *testing.T) {
	var tests = []struct {
		username string
		path     string
		result   bool
	}{
		{"foo", "/", false},
		{"foo", "/foo", true},
		{"foo", "/foo/", true},
		{"foo", "/foo/bar", true},
		{"foo", "/foobar", false},
	}

	for _, test := range tests {
		result := isUserPath(test.username, test.path)
		if result != test.result {
			t.Errorf("isUserPath(%q, %q) was incorrect, got: %v, want: %v.", test.username, test.path, result, test.result)
		}
	}
}
```

**Comment by Mebus** [comment]
Do you want to put this into a new PR?

**Comment by wojas** [comment]
@Mebus Feel free to just squash and push to this one.

**Comment by rawtaz** [comment]
Ping :) What's the current take on this?

**Comment by rawtaz** [comment]
@Mebus @wojas Ping :) What's the current take on this? Can it be merged so we get multi-tenancy in rest-server?

**Comment by Mebus** [comment]
Sorry, I don't have much time at the moment. I can look at it, but this will take some time.

Mebus

**Comment by zcalusic** [comment]
@Mebus, it would be great if you could spare some time to get this PR in order. It looks this feature has it use(r)s, and I'd be very glad to merge PR from you. Possibly before the next release, which is now overdue (I tried to follow one month release cycle, and failed at that).

OTOH, if you get into trouble (and we're all too busy, i can definitely attest to that 😢 ), say so, and maybe somebody might jump in.

**Comment by rawtaz** [comment]
I haven't delved into the code, but from the discussion above it seems to me that it's just a matter of implementing @wojas code?

**Comment by Mebus** [comment]
@zcalusic I updated my branch. Looks good so far :-)

**Comment by zcalusic** [comment]
Thank you @Mebus. I'm in the middle of another request, and will get back to this, as soon as I'm finished with the former.

In the meantime see if you can fix the conflict yourself. The code obviously evolved in the meantime, and there was a brand new release yesterday. Basically, you'd need to rebase the patch on the most recent commit in the master, so that it applies cleanly once again. GitHub suggests that it will require some manual intervention.

**Comment by Mebus** [comment]
I already pulled the current master into my branch and merged. I don't see any problems.

**Comment by zcalusic** [comment]
Ah, interesting. I had "Rebase and merge" selected and it reports issues, some GitHub pecularity. All the better, I'll check and probably merge it very soon.

File: mux.go [hunk 1]
```diff
@@ -14,16 +14,17 @@ import (
...
+	Log          string
+	CPUProfile   string
+	TLS          bool
+	TLSKey       string
+	TLSCert      string
+	AppendOnly   bool
+	PrivateRepos bool
+	Prometheus   bool
+	Debug        bool
 }{
```
Comment (zcalusic) [inline mux.go]: Please do not revert this recent change and move TLS bool back up. Even if it's a slight optimization, I intend to follow linters suggestion and practice good struct packing. Thanks.

**Comment by zcalusic** [comment]
OK, I went through the changes, and the bulk of functionality looks fine.

2 observations, the first I left as an review, this PR actually wants to revert one recent change: optimized layout of Config struct. Also you added several empty lines in various places, and deleted one line in one place. I didn't want to review line by line, but if you could remove those whitespace changes, it makes it easier to concentrate on what the patch really does.

If you could do those final fixes, and then it's ready to go in. And I also promise to release 0.9.7 soon, which will then have this nice feature in.

**Comment by fd0** [comment]
Uh, please do not merge the master branch back into a feature branch, that's a nightmare to understand later. Please rebase your changes onto the master branch, then it'll be **much** easier to understand what the commits do.

**Comment by Mebus** [comment]
Ok. Now I rebased to 0105f34a7310db67c1b30dbdfb0c2a601940219b. Does this help?

Mebus

**Comment by rawtaz** [comment]
@Mebus The latest commit on the master branch is https://github.com/restic/rest-server/commit/f99197dcf9e5f84bd3f9ebb40b9e611314845932 - I think that's what you should rebase your work onto?

**Comment by Mebus** [comment]
Can somebody tell me how to do this, please?

Mebus

**Comment by rawtaz** [comment]
Sure thing. The below assumes that the remote named 'upstream' is the one for this repository (I assume your own rest-server repository on GitHub is named 'origin'). Change as needed.

First, make sure you are on your working branch for this PR, e.g. `git checkout privaterepos`.

Next, make sure you have the latest commits from upstream using `git fetch upstream`.

Then rebase your current working branch onto upstream's master: `git rebase upstream/master`.

If there are any merge conflicts during the rebase, Git will tell you and also inform you that once you resolved those merge conflicts you can use `git rebase --continue` to continue completing there base. Here's a snippet from what Git outputs in those cases:

~~~
When you have resolved this problem run "git rebase --continue".
If you would prefer to skip this patch, instead run "git rebase --skip".
To check out the original branch and stop rebasing run "git rebase --abort".
~~~

Once this is done, you can update your PR by pushing to your remote branch on GitHub, e.g. `git push --force`, which should make sure that your PR is updated with the latest state of your privaterepos branch.

The Git Book has a good chapter on rebasing like this:  https://git-scm.com/book/en/v2/Git-Branching-Rebasing

**Comment by zcalusic** [comment]
Good job @Mebus & @wojas!

Let's get it in and test, test, test... Yeah, @rawtaz, I'm looking at you. ;)

**Comment by rawtaz** [comment]
@Mebus Nice work!

@zcalusic Thanks! I will *definitely* test this as soon as I get a chance. I have a new server waiting at a client, will use it for some real testing with 20 or so users in a few weeks.

**Comment by zcalusic** [comment]
Thanks @rawtaz, I appreciate it!

The 0.9.7 will be based on your positive review.

**Pull request merged**